pkg/controller/direct/tags/tagkey_controller.go

```diff
@@ -0,0 +1,353 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tags
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	krm "github.com/GoogleCloudPlatform/k8s-config-connector/apis/tags/v1beta1"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/config"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct/common"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct/directbase"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct/registry"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/structuredreporting"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/types/known/fieldmaskpb"
+
+	api "cloud.google.com/go/resourcemanager/apiv3"
+	pb "cloud.google.com/go/resourcemanager/apiv3/resourcemanagerpb"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func init() {
+	registry.RegisterModel(krm.TagsTagKeyGVK, NewTagsTagKeyModel)
+}
+
+func NewTagsTagKeyModel(ctx context.Context, config *config.ControllerConfig) (directbase.Model, error) {
+	return &TagsTagKeyModel{config: config}, nil
+}
+
+var _ directbase.Model = &TagsTagKeyModel{}
+
+type TagsTagKeyModel struct {
+	config *config.ControllerConfig
+}
+
+func (m *TagsTagKeyModel) AdapterForObject(ctx context.Context, reader client.Reader, u *unstructured.Unstructured) (directbase.Adapter, error) {
+	tagKeysClient, err := newTagKeysClient(ctx, m.config)
+	if err != nil {
+		return nil, err
+	}
+
+	obj := &krm.TagsTagKey{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &obj); err != nil {
+		return nil, fmt.Errorf("error converting to %T: %w", obj, err)
+	}
+
+	if err := common.NormalizeReferences(ctx, reader, obj, nil); err != nil {
+		return nil, fmt.Errorf("normalizing references: %w", err)
+	}
+
+	// normalize parent (sadly not a ref, for legacy reasons)
+	{
+		parent := direct.ValueOf(obj.Spec.Parent)
+		tokens := strings.Split(parent, "/")
+		if len(tokens) == 2 && tokens[0] == "projects" {
```

Comment:
NIT: len(tokens) == 2 needs to always be true. If we check that first (once) we could probably give a more targeted error and have less complicated conditional statements.